contract/contracts/expense-account/ExpenseAccountEIP712.sol

```diff
@@ -5,36 +5,106 @@ import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
     EIP712Upgradeable, 
     PausableUpgradeable {
 
-    struct BudgetLimit {
-        address approvedAddress;
+    using Address for address payable;
+    using ECDSA for bytes32;
+
+    /**
+     * @title BudgetData
+     * @notice Represents a specific budget constraint applied to withdrawals.
+     * @dev Combines a constraint type (`budgetType`) with its associated value.
+     * 
+     * @param budgetType The type of budget constraint to apply.
+     * @param value The value associated with the budget constraint:
+     * - For `TransactionsPerPeriod`: the maximum number of transactions allowed.
+     * - For `AmountPerPeriod`: the total allowed withdrawal amount within the period.
+     * - For `AmountPerTransaction`: the maximum amount per transaction.
+     */
+    struct BudgetData {
         BudgetType budgetType;
         uint256 value;
+    }
+
+    /**
+     * @title BudgetLimit
+     * @notice Represents a collection of budget constraints for an authorized transfer.
+     * @dev This struct is signed and verified using EIP-712, ensuring constraints are enforced.
+     * 
+     * @param approvedAddress The address authorized to perform withdrawals under these constraints.
+     * @param budgetData An array of budget constraints to apply to transfers.
+     * Each element in the array represents a specific `BudgetData` constraint.
+     * @param expiry The expiry timestamp for the budget constraints.
+     * After this timestamp, the constraints are no longer valid, and transfers are prohibited.
+     */
+    struct BudgetLimit {
+        address approvedAddress;
+        BudgetData[] budgetData;
         uint256 expiry;
     }
 
+    /**
+     * @title BudgetType
+     * @notice Defines the types of constraints that can be applied to withdrawals in the smart contract.
+     * @dev Used in conjunction with EIP-712 for signing and verifying constraints on withdrawals.
+     * 
+     * @param TransactionsPerPeriod Limits the number of transactions allowed within a specific time period.
+     * @param AmountPerPeriod Limits the total amount of funds that can be withdrawn within a specific time period.
+     * @param AmountPerTransaction Limits the maximum amount that can be withdrawn in a single transaction.
+     */
     enum BudgetType {
         TransactionsPerPeriod,
         AmountPerPeriod,
         AmountPerTransaction
     }
 
+    /// @dev Type signature for the BudgetData struct, used in EIP-712 encoding.
+    string private constant BUDGETDATA_TYPE = "BudgetData(uint8 budgetType,uint256 value)";
+
+    /// @dev Type signature for the BudgetLimit struct, used in EIP-712 encoding.
+    string private constant BUDGETLIMIT_TYPE = 
+        "BudgetLimit(address approvedAddress,BudgetData[] budgetData,uint256 expiry)";
+
+    /// @dev Typehash for the BudgetData struct, used in EIP-712 encoding.
+    bytes32 constant BUDGETDATA_TYPEHASH = 
+        keccak256(
+            abi.encodePacked(BUDGETDATA_TYPE)
+        );
+
```

Comment:
will u upgrade the contract implementation or u will redeploy the everything again? if so I think it can make storage clash between these new variables and `BUDGETLIMIT_TYPEHASH` also the other variables below. CMIIW

Reply:
I was going to redeploy everything since the expense account really hadn't started being used already so I wasn't checking for clashes etc.